contracts/test/UserRegistry.t.sol

```diff
@@ -0,0 +1,277 @@
+            address(0),
+            ROLE_SET_SUBREGISTRY | ROLE_SET_RESOLVER,
+            uint64(block.timestamp + 365 days)
+        );
+        
+        // Verify the domain was registered correctly
+        assertEq(proxy.ownerOf(tokenId), user1, "Domain should be owned by user1");
+        
+        // Verify roles were granted to the owner
+        bytes32 resource = proxy.getTokenIdResource(tokenId);
+        assertTrue(proxy.hasRoles(resource, ROLE_SET_SUBREGISTRY, user1), "User1 should have SET_SUBREGISTRY role");
+        assertTrue(proxy.hasRoles(resource, ROLE_SET_RESOLVER, user1), "User1 should have SET_RESOLVER role");
+        
+        // Verify the domain resolves correctly
+        assertEq(address(proxy.getSubregistry(label)), address(0), "Subregistry should be zero address");
+        assertEq(proxy.getResolver(label), address(0), "Resolver should be zero address");
+    }
+    
+    function test_domain_management() public {
+        // Register a domain
+        vm.prank(admin);
+        uint256 tokenId = proxy.register(
+            "mdtdomain",
+            user1,
+            IRegistry(address(0)),
+            address(0),
+            ROLE_SET_SUBREGISTRY | ROLE_SET_RESOLVER,
+            uint64(block.timestamp + 365 days)
+        );
+        
+        // User1 sets a resolver
+        address resolver = address(0x123);
+        vm.prank(user1);
+        proxy.setResolver(tokenId, resolver);
+        
+        // Verify resolver was set
+        assertEq(proxy.getResolver("mdtdomain"), resolver, "Resolver should be set");
+        
+        // User1 sets a subregistry
+        vm.prank(user1);
+        proxy.setSubregistry(tokenId, IRegistry(address(0x456)));
+        
+        // Verify subregistry was set
+        assertEq(address(proxy.getSubregistry("mdtdomain")), address(0x456), "Subregistry should be set");
+    }
+    
+    function test_role_management() public {
+        // Admin grants ROLE_REGISTRAR to user1
+        vm.prank(admin);
+        proxy.grantRootRoles(ROLE_REGISTRAR, user1);
+        
+        // Verify user1 has ROLE_REGISTRAR
+        assertTrue(proxy.hasRootRoles(ROLE_REGISTRAR, user1), "User1 should have registrar role");
+        
+        // User1 should be able to register domains now
+        vm.prank(user1);
+        uint256 tokenId = proxy.register(
+            "user1domain",
+            user2,
+            IRegistry(address(0)),
+            address(0),
+            ROLE_SET_SUBREGISTRY | ROLE_SET_RESOLVER,
+            uint64(block.timestamp + 365 days)
+        );
+        
+        // Verify registration was successful
+        assertEq(proxy.ownerOf(tokenId), user2, "Domain should be owned by user2");
+    }
+    
+    function test_Revert_unauthorized_registration() public {
+        // User1 tries to register a domain without ROLE_REGISTRAR
+        vm.prank(user1);
+        vm.expectRevert(); // Should revert because user1 doesn't have ROLE_REGISTRAR
```

Comment:
This should test for the exact reversion error so that we know it's the revert we expect.